### PR TITLE
Fix no_cache provider behavior

### PR DESCRIPTION
**Pull request by beldmit**
As we agreed today, this is a follow-up OpenSSL-only test to #26038 
Update: also the fix

Fixes: #26038

This tests modifies a NULL provider so it returns *no_cache equal to 1.
Surprisingly no other tests seem to fail.

##### Checklist

- [x] tests are added or updated

@levitte if you have any clues what could go wrong it would be great

CC @simo5

### CODE REVIEW

**Comment by beldmit** [comment]
@t8m thanks! Your hint about the null provider was very helpful

**Comment by beldmit** [comment]
I have heavily instrumented the code with debug prints and now I hopefully understand what's going on. If the provider with no_cache is activated, we create a temporary store for the methods provided by it. After that the methods from the default provider also go to this store (which seems wrong to me) and then on fetch this temporary store is not iterated. I think that for the default provider the store from the libctx should be used. I'll try to implement a fix tomorrow but I'm quite frustrated because of the callback architecture that makes things too difficult to debug

**Comment by beldmit** [comment]
All tests passed
Ready for review.

**Comment by paulidale** [comment]
Do we permanently want the null provider to be non-caching?

I cannot think of a problem with this -- it will create a negligible performance loss since it will be queried on fetch but return nothing.

**Comment by beldmit** [comment]
I don't see any problems with this

**Review by t8m — APPROVED** [review]

**Comment by beldmit** [comment]
https://github.com/openssl/openssl/commit/e1eafe8c87612a94552e9ad5df56c489cb6f0ff2 is the commit in question

**Comment by beldmit** [comment]
@t8m are you sure 3.0/3.1 are also problematic?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I do not think changing the null provider permanently to "no cache" makes sense - just for the purposes of testing. While I can't think that this would be too bad, I am not 100% certain that this has no effect in practice. We should use a separate test provider for this purpose.

**Comment by beldmit** [comment]
I've modified the p_test provider now.

@t8m @mattcaswell - could you please approve?

File: test/null-and-default.cnf
Comment (mattcaswell) [inline test/null-and-default.cnf:1]: The filename seems like it is no longer appropriate

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/null-and-default.cnf
Comment (beldmit) [reply test/null-and-default.cnf:1]: Fixed

Comment (beldmit) [reply test/null-and-default.cnf:1]: Thanks!

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
> @t8m are you sure 3.0/3.1 are also problematic?

Yes, https://github.com/openssl/openssl/pull/18153 was backported to 3.0 and is present in 3.1 as well.

File: test/recipes/20-test_nocache.t [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+use OpenSSL::Test::Utils;
+
+setup("test_nocache");
+
+plan tests => 4;
+
+ok(run(app(["openssl", "list", "-mac-algorithms"],
+        stdout => "listout.txt")),
+"List mac algorithms - default configuration");
+open DATA, "listout.txt";
```
Comment (shahsb) [inline test/recipes/20-test_nocache.t:22]: What if unable to open listout.txt? 

Could we add a check here?

File: test/recipes/20-test_nocache.t [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+my @match = grep /MAC/, <DATA>;
+close DATA;
+ok(scalar @match > 1 ? 1 : 0, "Several algorithms are listed - default configuration");
+
+$ENV{OPENSSL_CONF} = bldtop_file("test", "nocache-and-default.cnf");
+ok(run(app(["openssl", "list", "-mac-algorithms"],
+        stdout => "listout.txt")),
+"List mac algorithms");
+open DATA, "listout.txt";
```
Comment (shahsb) [inline test/recipes/20-test_nocache.t:31]: Same here:
What if unable to open listout.txt? 

Could we add a check here?

**Review by shahsb — CHANGES_REQUESTED** [review]

File: test/recipes/20-test_nocache.t (hunk 1 above)
Comment (beldmit) [reply test/recipes/20-test_nocache.t:22]: No, it doesn't make sense. If our tests can't read or write to the folder they are run in, it means problems with setup

File: test/recipes/20-test_nocache.t (hunk 2 above)
Comment (beldmit) [reply test/recipes/20-test_nocache.t:31]: Same here

**Comment by beldmit** [comment]
Merged to 3.2+
Will raise a separate PR to 3.0 to avoid merge conflicts

Thank you for review!

**Pull request closed without merging**